### PR TITLE
Added IP Validation with spinner field

### PR DESCRIPTION
**Pull request by yoshie902a**
Will create a hidden field with hash of ip, date, and secret key, then validates it  This prevents attacks from different ip addresses, and validates that the date session data was not tampered with.

Feel free to refactor and improve.

### CODE REVIEW

**Comment by markets** [comment]
Hi @yoshie902a and thanks a lot for opening this discussion and contributing to `invisible_captcha`. As I commented in the original issue, in principle I like the idea of adding some kind of "spinner" to this gem, with the aim to add more barriers to bots. Let add some comments to the implementation.

File: README.md [hunk 1]
```diff
@@ -120,13 +122,37 @@ InvisibleCaptcha.setup do |config|
...
 
+#### IP Enabled Setup
+
+View code:
+
+```erb
+<%= form_for(@topic) do |f| %>
+  <%= f.invisible_captcha :subtitle, { timestamp: @invisible_captcha_values.timestamp, spinner_value: @invisible_captcha_values.spinner_value } %>
+  <!-- or -->
+  <%= invisible_captcha :subtitle, :topic, { timestamp: @invisible_captcha_values.timestamp, spinner_value: @invisible_captcha_values.spinner_value } %>
```
Comment (markets) [inline README.md:142]: Ideally this shouldn't be set by the user, I think we can do it automatically under the hood if the `ip_check` is enabled

File: README.md [hunk 2]
```diff
@@ -120,13 +122,37 @@ InvisibleCaptcha.setup do |config|
...
+<% end %>
+```
+
+Controller code:
+
+```ruby
+class TopicsController < ApplicationController
+  before_action :invisible_captcha_values, only: %i[new create update]
```
Comment (markets) [inline README.md:150]: Similar to https://github.com/markets/invisible_captcha/pull/85/files#r573288133, would be ideal to do this without the user configuring/adding more lines

File: lib/invisible_captcha.rb [hunk 3]
```diff
@@ -36,6 +39,12 @@ def init!
       # If enabled, you should call anywhere in of your layout the following helper, to inject the honeypot styles:
       #  <%= invisible_captcha_styles %>
       self.injectable_styles = false
+      
+      # IP check enabled by default
+      self.ip_enabled = true
+      
+      # A secret key setup in config settings. 'rake secret' will give you a good one.
+      self.secret = Digest::MD5.hexdigest("make_sure_to_set_your_secret_key_in_config_settings_do_not_use_default")
```
Comment (markets) [inline lib/invisible_captcha.rb:47]: Maybe be can default to something like Devise does:

https://github.com/heartcombo/devise/blob/45b831c4ea5a35914037bd27fe88b76d7b3683a4/lib/generators/templates/devise.rb#L17

File: lib/invisible_captcha/controller_ext.rb [hunk 4]
```diff
@@ -17,9 +18,25 @@ def invisible_captcha(options = {})
...
+
+    def spinner_value
+       @spinner_value ||= InvisibleCaptcha.encode("#{invisible_captcha_timestamp}-#{request.remote_ip}")
+    end
 
     def detect_spam(options = {})
-      if timestamp_spam?(options)
+      if ip_spam?(options)
+        on_spam(options)
+      elsif timestamp_spam?(options)
```
Comment (markets) [inline lib/invisible_captcha/controller_ext.rb:39]: I think we should keep first the timestamp check

File: lib/invisible_captcha/controller_ext.rb [hunk 5]
```diff
@@ -74,11 +91,25 @@ def timestamp_spam?(options = {})
 
       return false
     end
-
-    def honeypot_spam?(options = {})
+    
+    def ip_spam?(options ={})
```
Comment (markets) [inline lib/invisible_captcha/controller_ext.rb:95]: I think we can simplify a bit this method if we make the changes suggested in https://github.com/markets/invisible_captcha/pull/85/files#r573288133

File: lib/invisible_captcha/view_helpers.rb [hunk 6]
```diff
@@ -1,4 +1,4 @@
-# frozen_string_literal: true
+#frozen_string_literal: true
```
Comment (markets) [inline lib/invisible_captcha/view_helpers.rb:1]: 🙏🏼  let's try to avoid unnecessary changes please, like this one or some trailing whitespaces

File: README.md (hunk 1 above)
Comment (yoshie902a) [reply README.md:142]: I would agree, but I ran into a problem with getting the helper_method working..

I posted this...  https://stackoverflow.com/questions/66112686/railtie-and-undefined-method-helper-method-whats-the-deal

If you have any ideas, please let me know.

File: lib/invisible_captcha.rb (hunk 3 above)
Comment (yoshie902a) [reply lib/invisible_captcha.rb:47]: Accessing the key is different for each rails version, so it was easier and cleaner to have the user set one.

File: lib/invisible_captcha/controller_ext.rb (hunk 4 above)
Comment (yoshie902a) [reply lib/invisible_captcha/controller_ext.rb:39]: changed the order.

File: README.md (hunk 2 above)
Comment (yoshie902a) [reply README.md:150]: I agree, made change.

File: lib/invisible_captcha/view_helpers.rb (hunk 6 above)
Comment (yoshie902a) [reply lib/invisible_captcha/view_helpers.rb:1]: Sorry, unintentional

File: lib/invisible_captcha.rb (hunk 3 above)
Comment (yoshie902a) [reply lib/invisible_captcha.rb]: I added the code Devise used for secret key finder, but the way the gem loads, Rails.application does not load before setting the value. @markets can take a look as to the best way to get the loading order correct. This is the first gem I have contributed to, so not sure of the best structure to do this.

File: README.md (hunk 1 above)
Comment (markets) [reply README.md:143]: Ok! sorry I have very limited time this week, I'll try to take a look during the weekend. But ideally, I'd like to remove all public API related to this new spinner thing. Probably, one thing we can explore is to make all this machinery automatically using the `invisible_captcha` view helper.

File: lib/invisible_captcha.rb (hunk 3 above)
Comment (markets) [reply lib/invisible_captcha.rb]: but the line I copied above is just:

```
config.secret_key = '<%= SecureRandom.hex(64) %>'
```

nothing related to Rails, but IMO better than `Digest::MD5.hexdigest("make_sure_to_set_your_secret_key_in_config_settings_do_not_use_default")` which requires the user to change it.

File: README.md [hunk 7]
```diff
@@ -120,13 +122,40 @@ InvisibleCaptcha.setup do |config|
...
+```
+
+Controller code:
+
+```ruby
+class TopicsController < ApplicationController
+  # By default, invisible_captcha sets instance variabls for all actions,
+  # but does not run detect_spam for actions specified.
+  _
+  invisible_captcha only: [:new, :create, :update], honeypot: :subtitle
```
Comment (markets) [inline README.md:154]: sorry, this is really weird API and can potentially break current people code. I don't want to break people app with sensible stuff like forms, sign-ups, ...

I still think we should find a way to make all this new "spinner" thing under the hood without the users updating and changing all current `invisible_captcha` occurrences in their apps.

File: lib/invisible_captcha.rb [hunk 8]
```diff
@@ -1,12 +1,15 @@
 # frozen_string_literal: true
-
+require 'rails'
+require 'responders'
```
Comment (markets) [inline lib/invisible_captcha.rb:3]: those 2 requires are unnecessary, AFAIK `responders` is not a Rails dependency, so this will crash as you can see in the CI https://travis-ci.com/github/markets/invisible_captcha/jobs/481937373#L320:

```
Failure/Error: require 'responders'
```

File: lib/invisible_captcha.rb [hunk 9]
```diff
@@ -36,6 +42,16 @@ def init!
...
+      
+      # IP check enabled by default
+      self.ip_enabled = true
+      
+      # A secret key setup in config settings. 'rake secret' will give you a good one.
+      # Rails.application not currently loading. Need to change the way this loads.
+      self.secret = 'Secret' # InvisibleCaptcha::SecretKeyFinder.new(Rails.application).find
+   
+      # Actions that exclude running the detect_spam method, usd mainly for view actons
+      self.exclude_detect_actions = ['new', 'edit']
```
Comment (markets) [inline lib/invisible_captcha.rb:54]: as commented above, this is a bit confusing API from my point of view... what about invisible_captcha occurrences for non-rest resources?

File: lib/invisible_captcha/controller_ext.rb [hunk 10]
```diff
@@ -70,11 +99,28 @@ def timestamp_spam?(options = {})
...
       honeypot = options[:honeypot]
       scope    = options[:scope] || controller_name.singularize
 
+      if InvisibleCaptcha.ip_enabled
+        if params[:spinner] == spinner_value || (params[scope] && params[scope][:spinner] == spinner_value)
```
Comment (markets) [inline lib/invisible_captcha/controller_ext.rb:108]: sorry for insisting on this, but all these things can be simplified if we found a way to make all this work under the hood (as we do for example with the time sensitive check, final users only need to enable/disable)

**Comment by markets** [comment]
Hi @yoshie902a I added some comments. 

I actually have very limited time to maintain all my OSS projects, so I only want to merge things that I can maintain in the long run (and have good tests 👌🏼 , of course). The current path we are taking in this branch is a bit weird to be honest. As exposed several times, I think we should implement this in a more transparent way for the user (just a flag: ON/OFF).

As I also commented in my first comment https://github.com/markets/invisible_captcha/pull/85#issuecomment-776281486, if you want to push new things, please rebase from master (I simplified the Rails compatibility matrix and moved new travis infra).

Best,
Marc

File: lib/invisible_captcha/controller_ext.rb (hunk 10 above)
Comment (yoshie902a) [reply lib/invisible_captcha/controller_ext.rb:108]: I completely agree with everything you are saying.  The main difference between the time sensitive check and the spinner is that we need to set values from the controller and make them available to the views.  However, some need to set the values and some actions need to detect them.  If the helper_method issue was solved, we would not need the before_action and I think the API would be left alone.

**Comment by yoshie902a** [comment]
If you have time to solve the helper_method issue, I can refactor my changes to fix it the way you want it.  Also, I thought I rebased. Do you not see it?

**Comment by yoshie902a** [comment]
I think I solved the helper_method issue. Working on fixing the rest. If you like the implementation after I push, then I will write the testing for it.

**Comment by yoshie902a** [comment]
I am much happier with this code, I agreed with all your comments.  Once I fixed the helper_method issue, it was much easier to structure the code.  If you are happy with the current changes, I will write the tests for them.

**Comment by markets** [comment]
Hi @yoshie902a I'll make a deep look/review during the weekend, but I like much more this new path we are taking. Thanks!

**Comment by yoshie902a** [comment]
**When you review this weekend**: Your testing style is a bit different than myself.  To make sure the code is maintainable the way you like, maybe you can make your more in-depth code review by adding the needed tests?  That way, you know it's tested the way you like and it checks everything you prefer. If not, I can try, but I unfortunately may not have time to redo it.

**Comment by markets** [comment]
ok @yoshie902a no prob! but current CI doesn't pass on any combination of Ruby/Rails:

Captura de pantalla 2021-02-11 a las 19 16 49

So, at least we need a green build just to be sure we are not breaking things with these changes.

**Comment by yoshie902a** [comment]
Tests are passing, by adding the two helpers and making ip_enabled false.  If ip_enabled is true, behavior changes and tests need to be added/changed.

**Comment by markets** [comment]
Hi @yoshie902a I pushed a new branch (https://github.com/markets/invisible_captcha/compare/ip_validation) with some clean-up. The branch is still WIP, I want to update the docs a bit and test the branch in some of my apps. Thanks!

**Comment by yoshie902a** [comment]
Thanks! let me know, when you merge it!

**Comment by markets** [comment]
@yoshie902a going to close this one, new work in 👉🏼  #89
